### ISSUE TITLE
False positive with `prefer-const` in delayed assignment

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#8871]
**Tell us about your environment**

* **ESLint Version:** 4.1.1
* **Node Version:** 8.0.0
* **Yarn Version:** 0.24.6

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:**

```
---
parserOptions:
    ecmaVersion: 6
rules:
    prefer-const: error
```

**What did you do? Please include the actual source code causing the issue.**

```js
let foo;
bar();
foo = 1;
```

**What did you expect to happen?**

No errors. (The assignment isn't until later.)

**What actually happened? Please include the actual, raw output from ESLint.**

```
3:1 - 'foo' is never reassigned. Use 'const' instead. (prefer-const)
```

**Comment by isiahmeadows** [comment]
For context, I actually encountered it in code that looks closer to this in practice:

```js
let foo;
let bar = somethingAsync(() => resolve(foo));
foo = getValue(bar);
```

**Comment by isiahmeadows** [comment]
And also, I tested the minimal repro online just a few minutes ago.

**Comment by gyandeeps** [comment]
You have to activate `ignoreReadBeforeAssign` options.

`/*eslint prefer-const: ["error", {"ignoreReadBeforeAssign": true}]*/`

Docs: http://eslint.org/docs/rules/prefer-const#ignorereadbeforeassign

**Comment by isiahmeadows** [comment]
Oh okay.
